### PR TITLE
Finish data annotator view implementation

### LINKED ISSUE #438: [Feature Request]: Basic Datamap implementation
### Contact Details

_No response_

### Is your feature request related to a problem?

ARCtrl is done implementing datamap into model.

### Describe the solution you would like!

Either we implement it as extra file, or bonus worksheet. Download as worksheet or workbook.

- [x] display datamap in Swate as named sheet (similiar to metadata)
- [x] #465
- [x] #466
- [x] #467
- [x] #468

### PR DESCRIPTION
**Pull request by Freymaurer**
Closes #438

Adds data annotator view supporting tabular data files (.csv, .tsv) ✨ 
This feature can be used on tables, as well as data map!
Uses a lazy load table, which is tested for 100 000 rows with only minor performance issues!

> [!IMPORTANT]
> In the future we might benefit from range selection. Such as cell-1,1 to cell 2,5. Same for columns and rows.

* Usable from widget or sidebar page
image

* Uploading a file will open a modal:
DataAnnotatorView

* a) Can be used to update the separator used to split lines in columns. Swate by default tries to guess the separator based on file extension (e.g. `,` for .csv)
* b) Can be toggled to make the first row header or just another value row.
* c) Used to declare where to insert the data when **Submit**ting. Hover will show more information. 

```
Autodetect -> Creates missing Input or Output column, if both exist submit will fail!
Input -> Create Input column, will overwrite!
Output -> Create Output column, will overwrite!
```

* d) Click on column header to select full column
* e) Click on row header to select full row
* f) Click on cell to select cell

> [!IMPORTANT]
> All fields with `+` will create one data selector!

The output created follows [specification](https://github.com/nfdi4plants/ARC-specification/blob/main/ISA-XLSX.md#inputs-and-outputs):

image

@Brilator @HLWeil @muehlhaus @ZimmerD

### CODE REVIEW

File: src/Client/MainComponents/SpreadsheetView/DataMap.fs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            )    
+    ]
+
+let private BodyRows (state:Set<int>) (model:Model) (dispatch: Msg -> unit) =
+    [|
+        for rowInd in 0 .. model.SpreadsheetModel.DataMapOrDefault.RowCount-1 do
+            yield BodyRow rowInd state model dispatch 
+    |]
+
+let private HeaderRow (state:Set<int>) setState (model:Model) (dispatch: Msg -> unit) =
```
Comment (Etschbeijer) [inline src/Client/MainComponents/SpreadsheetView/DataMap.fs:88]: Same as above, naming of the function is not helpful

File: src/Client/MainComponents/SpreadsheetView/DataMap.fs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            (
+                columnIndex,
+                rowIndex,
+                state,
+                model,
+                dispatch
+            )    
+    ]
+
+let private BodyRows (state:Set<int>) (model:Model) (dispatch: Msg -> unit) =
```
Comment (Etschbeijer) [inline src/Client/MainComponents/SpreadsheetView/DataMap.fs:82]: Same as above, naming of the function is not helpful

File: src/Client/SidebarComponents/Navbar.fs [hunk 3]
```diff
@@ -222,14 +222,21 @@ let NavbarComponent (model : Model) (dispatch : Msg -> unit) (sidebarsize: Model
                                     )
                                     prop.text Routing.Route.Info.toStringRdbl
                                 ]
+                                Bulma.navbarItem.a [
+                                    prop.onClick (fun e ->
```
Comment (Etschbeijer) [inline src/Client/SidebarComponents/Navbar.fs:226]: e is never used, use _ instead

File: src/Client/Spreadsheet/Controller/BuildingBlocks.fs [hunk 4]
```diff
@@ -63,6 +63,30 @@ let addBuildingBlocks(newColumns: CompositeColumn []) (state: Spreadsheet.Model)
...
+let addDataAnnotation (data: {| fragmentSelectors: string []; fileName: string; fileType: string; targetColumn: DataAnnotator.TargetColumn |}) (state: Spreadsheet.Model) : Spreadsheet.Model =
+    let tryIfNone() =
+        match state.ActiveTable.TryGetInputColumn(), state.ActiveTable.TryGetOutputColumn() with
+        | Some _, None
+        | None, None        -> CompositeHeader.Output IOType.Data
+        | None, Some _      -> CompositeHeader.Input IOType.Data
+        | Some _, Some _    -> failwith "Both Input and Output columns exist and no target column was specified"
+    let newHeader =
+        match data.targetColumn with
+        | DataAnnotator.TargetColumn.Input -> CompositeHeader.Input IOType.Data
```
Comment (Etschbeijer) [inline src/Client/Spreadsheet/Controller/BuildingBlocks.fs:75]: allign the output and -> please, makes it easier to read

File: src/Client/Helper.fs [hunk 5]
```diff
@@ -87,6 +87,38 @@ let debouncel<'T> (storage:DebounceStorage) (key: string) (timeout: int) (setLoa
...
 
+let debouncemin (fn: 'a -> unit, timeout: int) =
+    let mutable id = null: string
+    fun (arg: 'a) ->
+        match id with
+        | null -> ()
+        | id -> Fable.Core.JS.clearTimeout (int id)
+        let timeoutId = 
+            Fable.Core.JS.setTimeout
+                (fun () ->
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/Helper.fs:98]: Use `()` when input is actually unit. Use `_` when lambda argument is not unit and not used 👍

[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/Helper.fs:98]: In this specific case the `()` helps for type inference

File: src/Client/MainComponents/AddRows.fs [hunk 6]
```diff
@@ -13,7 +13,7 @@ let Main (dispatch: Messages.Msg -> unit) =
     Html.div [
         prop.id "ExpandTable"
         prop.style [
-            style.flexGrow 1; style.justifyContent.center; style.display.inheritFromParent; style.padding(length.rem 1)
+            style.flexGrow 0; style.justifyContent.center; style.display.inheritFromParent; style.padding(length.rem 1)
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/MainComponents/AddRows.fs:16]: Previously i wanted the "AddRows" Element to be below the table. With the change to the lazy load table i moved the element to be fixed above the navigation tabs. `flexGrow 0` equals `do not grow above your content size`

[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/MainComponents/AddRows.fs:16]: image

File: src/Client/MainComponents/Cells.fs [hunk 7]
```diff
@@ -245,30 +266,33 @@ type Cell =
...
-                let options = createEmpty<Browser.Types.ScrollIntoViewOptions>
-                options.behavior <- Browser.Types.ScrollBehavior.Auto
-                options.``inline`` <- Browser.Types.ScrollAlignment.Nearest
-                options.block <- Browser.Types.ScrollAlignment.Nearest
-                if ref.current.IsSome then ref.current.Value.scrollIntoView(options)), 
-                [|box isSelected|]
-        )
+        let cellId = Controller.Cells.mkCellId columnIndex rowIndex state
+        let ref = React.useElementRef()
+        //React.useEffect((fun _ ->
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/MainComponents/Cells.fs:277]: Not sure, i am not 100% happy with the scroll logic. Previously the code in this comment made the browser scroll zu the cell whenever it was updated to selected with arrow key navigation. 

I had some issues (and still have some issues) with this logic and the lazy load table. But for now i updated the logic to use `focus` (which is also the reason for `prop.tabIndex 0`; required to be focused).

For now i would leave the comment to reenable the logic when working for a permanent clean solution

File: src/Client/MainComponents/Metadata/Investigation.fs [hunk 8]
```diff
@@ -12,79 +12,84 @@ open Shared
...
+                )
+                FormComponents.OntologySourceReferencesInput(
+                    Array.ofSeq inv.OntologySourceReferences,
+                    "Ontology Source References",
+                    (fun oas -> 
+                        inv.OntologySourceReferences <- ResizeArray oas
+                        inv |> Investigation |> Spreadsheet.UpdateArcFile |> SpreadsheetMsg |> dispatch)
+                )
+                //FormComponents.TextInputs(
+                //    Array.ofSeq inv.RegisteredStudyIdentifiers,
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/MainComponents/Metadata/Investigation.fs:81]: Might be relevant in the near future, if this comment stays there for some months without negative user feedback we can delete it. Will add more info above

File: src/Client/MainComponents/SpreadsheetView/DataMap.fs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                Cell.HeaderTAN(columnIndex, header, state, setState, model, dispatch)
+        elif header.IsDataColumn then
+            Cell.HeaderDataSelector(columnIndex, header, state, setState, model, dispatch)
+            Cell.HeaderDataFormat(columnIndex, header, state, setState, model, dispatch)
+            Cell.HeaderDataSelectorFormat(columnIndex, header, state, setState, model, dispatch)
+        else
+            ()
+    ]
+
+let private BodyRow (rowIndex: int) (state:Set<int>) (model:Model) (dispatch: Msg -> unit) =
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/MainComponents/SpreadsheetView/DataMap.fs:59]: I'm not sure exactly what your point is. 😅

File: src/Client/MainComponents/SpreadsheetView/DataMap.fs (hunk 1 above)
Comment (Freymaurer) [reply src/Client/MainComponents/SpreadsheetView/DataMap.fs:88]: Same as above but rowIndex is not needed as we specifically want to iterate headers

File: src/Client/MainComponents/SpreadsheetView/DataMap.fs (hunk 2 above)
Comment (Freymaurer) [reply src/Client/MainComponents/SpreadsheetView/DataMap.fs:82]: Explained in comment for `BodyRow`

File: src/Client/Spreadsheet/Controller/BuildingBlocks.fs (hunk 4 above)
Comment (Freymaurer) [reply src/Client/Spreadsheet/Controller/BuildingBlocks.fs:75]: Yeah another reason to setup fornax. A few more and i get to it! 😄

File: src/Client/LocalStorage/Widgets.fs [hunk 10]
```diff
@@ -20,6 +20,7 @@ open Fable.SimpleJson
 let [<Literal>] BuildingBlockWidgets = "BuildingBlock"
 let [<Literal>] TemplatesWidgets = "Templates"
 let [<Literal>] FilePickerWidgets = "FilerPicker"
+let [<Literal>] DataAnnotatorWidgets = "DataAnnotator"
```
[the remark this answers is not in the document]
Comment (Freymaurer) [reply src/Client/LocalStorage/Widgets.fs]: Sort of, i am not really a fan of whitespace

File: src/Client/MainComponents/SpreadsheetView/DataMap.fs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Etschbeijer) [reply src/Client/MainComponents/SpreadsheetView/DataMap.fs:59]: Yes, I understood that so far but the naming it is named like a property and not what it is doing.
IterateColumns for BodyRow could be better suitable and IterateRows for BodyRows.

File: src/Client/MainComponents/SpreadsheetView/DataMap.fs (hunk 1 above)
Comment (Etschbeijer) [reply src/Client/MainComponents/SpreadsheetView/DataMap.fs:88]: see above

File: src/Client/MainComponents/SpreadsheetView/DataMap.fs (hunk 2 above)
Comment (Etschbeijer) [reply src/Client/MainComponents/SpreadsheetView/DataMap.fs:82]: see above

**Review by Etschbeijer — APPROVED** [review]

**Pull request merged**